### PR TITLE
docs: add information about @storybook/preview-api useArgs hook

### PR DESCRIPTION
**Pull request by jonthenerd**
Provide example for those migrating from v6 to v7 and were previously using the @storybook/client-api useArgs hook

Deals with (but likely isn't 100% relevant to) #12006 

## What I did

Added documentation to the seemingly relevant place with information about how to use the useArgs hook from within a story

## How to test

N/A these are docs. However, it would be best to test this within an actual component to ensure the docs are right. I have within my own v7 storybook.

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [X] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by jonthenerd** [comment]
No longer causes issue with eslint. However, another user that tried my solution reported an error that I can't reproduce and am not sure where it's coming from. It is not happening on my machine with Storybook v7.0.11 and addons/eslint config. See https://github.com/storybookjs/storybook/issues/12006#issuecomment-1551524737

The mentioned error seems like something Storybook itself is generating. Would be good to figure out how to ensure that doesn't generate before merging these helpful docs.

**Comment by jonthenerd** [comment]
Aligned code block with answer in #22593 . This PR is good to go for review :)

**Review by akashchouhan16 — APPROVED** [review]
🚀

File: docs/writing-stories/args.md
Comment (jonniebigodes) [inline docs/writing-stories/args.md:244]: @jonthenerd, if you're with it, could you update your pull request and factor in the following:
- Create a TypeScript snippet file inside the react folder of the documentation
- Do the same for a JavaScript variation
- Replace the code block here with reference to the filenames as follows:

```md
<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'react/this-pr-example-snippet.js.mdx',
    'react/this-pr-example-snippet.ts.mdx',
  ]}
/>

<!-- prettier-ignore-end -->
```

- If possible, run the `yarn pretty-docs` task on the monorepo so that the files run through prettier to avoid inconsistencies, typos, etc.

**Review by jonniebigodes — CHANGES_REQUESTED** [review]
@jonthenerd, thanks for taking the time to put together this pull request and improve the documentation. Appreciate it. I left one item for you to look into.

Let me know once you've addressed it and we'll take it from there.

Have a great week.

Stay safe

File: docs/writing-stories/args.md
Comment (jonthenerd) [reply docs/writing-stories/args.md]: @jonniebigodes I believe this is ready now. Thanks!

Comment (jonthenerd) [reply docs/writing-stories/args.md]: @jonniebigodes did my changes 3 weeks ago meet your expectations?

Comment (jonniebigodes) [reply docs/writing-stories/args.md]: @jonthenerd sorry for the radio silence on this. I'll take another look tomorrow morning my time and follow up with you on this. Sounds good?

File: docs/snippets/react/page-story-args-within-story.js.mdx
Comment (jonniebigodes) [inline docs/snippets/react/page-story-args-within-story.js.mdx:21]: @jonthenerd are you ok with simplifying this to the following and applying it to the TS versions (e.g., pre-4.9 and 4.9:

```js
/**
 * 👇 To avoid linting issues, it is recommended to use a function with a capitalized name or an arrow function
 */
 ```

File: docs/writing-stories/args.md
Comment (jonniebigodes) [inline docs/writing-stories/args.md:223]: @jonthenerd, what do you think about rephrasing it to:

```md
Interactive components often need to be controlled by their containing component or page to respond to events, modify their state and reflect those changes in the UI. For example, when a user toggles a switch component, the switch should be checked, and the arg shown in Storybook should reflect the change. To enable this, you can use the [`useArgs`](../addons/addons-api.md#useargs) API exported by `@storybook/preview-api`:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'react/page-story-args-within-story.js.mdx', 
    'react/page-story-args-within-story.ts.mdx',
  ]}
/>

<!-- prettier-ignore-end -->
```

File: docs/snippets/react/page-story-args-within-story.ts.mdx
Comment (jonniebigodes) [inline docs/snippets/react/page-story-args-within-story.ts.mdx:1]: @jonthenerd sorry for not mentioning it earlier, but could you also provide an example in the form of a snippet aimed at TypeScript 4.9? You don't need to link it, as it will be automatically picked up when the documentation is published; you can name it as `page-story-args-within-story.ts-4-9.mdx` and place it in the same folder as this example (e.g., `react`).

**Review by jonniebigodes — CHANGES_REQUESTED** [review]
@jonthenerd left a couple of items for you to look at when you're able. Let me know once you've addressed them and we'll take it from there. Sounds good?

**Comment by jonthenerd** [comment]
Working on this again. This is NOT ready for review yet. Will update when it is.

File: docs/snippets/react/page-story-args-within-story.js.mdx
Comment (jonthenerd) [reply docs/snippets/react/page-story-args-within-story.js.mdx]: @jonniebigodes I simplified it, but didn't keep it exactly as you provided as it read as if using an arrow function would pass linting (which it would not). I also added the typescript 4.9 version. This should be ready for re-review.

**Comment by jonniebigodes** [comment]
@jonthenerd no problem whatsoever; ping me when you're ready, and I'll be more than glad to take another look at it. Sounds good?

File: docs/writing-stories/args.md
Comment (jonthenerd) [reply docs/writing-stories/args.md]: Made the change to the text as requested. However, the link is to a page describing the `manager-api` and we're talking here about using the `preview-api`. Is that okay?

File: docs/snippets/react/page-story-args-within-story.ts.mdx
Comment (jonthenerd) [reply docs/snippets/react/page-story-args-within-story.ts.mdx:1]: done!

**Comment by jonthenerd** [comment]
@jonniebigodes think it's ready to go. Give it a whirl please :)

File: docs/writing-stories/args.md
Comment (jonniebigodes) [reply docs/writing-stories/args.md]: Sorry for the confusion here, but we shouldn't be pointing to an example aimed at the `manager-api` as it will introduce a misconception. If you don't mind, could you remove the link for now? I'll take a more thorough look tomorrow, and if nothing arises we can get this merged in.

Comment (jonthenerd) [reply docs/writing-stories/args.md]: Figured as much. Removed the link. Ready to go!

File: docs/writing-stories/args.md
Comment (jonniebigodes) [inline docs/writing-stories/args.md:216]: @jonthenerd, one final thing, as the underlying issue relates to React and to avoid confusion, we have a mechanism in place that allows the documentation to render for specific frameworks conditionally. Would you mind just wrapping this with the following `<IfRenderer renderer='react'> ....</IfRenderer>` Turning this into:

```markdown
<IfRenderer renderer='react'>

## Setting args from within a story


Interactive components often need to be controlled by their containing component or page to respond to events, modify their state and reflect those changes in the UI. For example, when a user toggles a switch component, the switch should be checked, and the arg shown in Storybook should reflect the change. To enable this, you can use the `useArgs` API exported by `@storybook/preview-api`:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[    
    'react/page-story-args-within-story.js.mdx',
    'react/page-story-args-within-story.ts.mdx'
  ]}
/>

<!-- prettier-ignore-end -->

</IfRenderer>

``` 


If you have any questions, see how it's done on the [install page](https://github.com/storybookjs/storybook/blob/next/docs/get-started/install.md).

Comment (jonthenerd) [reply docs/writing-stories/args.md:218]: @jonniebigodes added the IFRenderer as requested. Hopefully this is the last bit? Believe this is ready for re-review.

**Comment by jonthenerd** [comment]
@jonniebigodes this is ready for re-review

**Comment by jonthenerd** [comment]
@jonniebigodes , perhaps you're on vacation? @akashchouhan16 is there someone else who should review?

**Comment by jonniebigodes** [comment]
@jonthenerd sorry for the radio silence. I was away for some time off, but I haven't forgotten about this. If you don't mind, I'll check it on Monday and if nothing arises, I'll gladly merge it.

Hope you have a fantastic weekend.

Stay safe

**Comment by jonthenerd** [comment]
> @jonthenerd sorry for the radio silence. I was away for some time off, but I haven't forgotten about this. If you don't mind, I'll check it on Monday and if nothing arises, I'll gladly merge it.
> 
> Hope you have a fantastic weekend.
> 
> Stay safe

Sounds good @jonniebigodes . Hope your time off was restful.

**Review by jonniebigodes — APPROVED** [review]
@jonthenerd, as promised, I've checked the pull request, and all is good on my end. Once again, thanks for taking the time to help us improve the documentation with this fantastic contribution. Appreciate it 🙏 ! 

I'll get it in once the checklist clears.

Hope you have a fantastic week.

Stay safe

**Pull request merged**
